### ISSUE TITLE
No need to wait after Wire.requestFrom().

### ISSUE THREAD

**Issue by Koepel** [UFABCRocketDesign/Embedded_System#1]
In the file "Tupa/Initial/Classes.cpp" you may remove all these lines after a `Wire.requestFrom()`:
```
unsigned long temp = micros();
while (Wire.available() < 22)
{
  if (temp + 10 < micros())break;
}
```
That code is not doing something.
Explanation: [Common-mistakes#1](https://github.com/Koepel/How-to-use-the-Arduino-Wire-library/wiki/Common-mistakes#1)

This code below updates the `X`, `Y` and `Z` only if the data was actually received:
```
Wire.requestFrom(address, (uint8_t)6);
if (6 <= Wire.available())
{
  X = Wire.read() << 8 | Wire.read();		//X_msb << 8 | X_lsb
  Z = Wire.read() << 8 | Wire.read();		//Z_msb << 8 | Z_lsb
  Y = Wire.read() << 8 | Wire.read();		//Y_msb << 8 | Y_lsb
}
```
But it does not show that the data is signed 16-bit integers. You could write it also like this:
```
Wire.requestFrom(address, (uint8_t)6);
if (Wire.available() == 6)
{
  int16_t x_data = Wire.read() << 8 | Wire.read();      //X_msb << 8 | X_lsb
  int16_t z_data = Wire.read() << 8 | Wire.read();      //Z_msb << 8 | Z_lsb
  int16_t y_data = Wire.read() << 8 | Wire.read();      //Y_msb << 8 | Y_lsb
  X = float(x_data);
  Z = float(z_data);
  Y = float(y_data);
}

```

**Comment by HeckRodSav** [comment]
My main concern is related to the case where the sensors are disconnected, the system can not stop working (it goes into a state of emergency). Would it still work?

In the other I can simply do:
```cpp
Wire.requestFrom(address, (uint8_t)6);
if (6 <= Wire.available())
{
  X = float(Wire.read() << 8 | Wire.read());		//X_msb << 8 | X_lsb
  Z = float(Wire.read() << 8 | Wire.read());		//Z_msb << 8 | Z_lsb
  Y = float(Wire.read() << 8 | Wire.read());		//Y_msb << 8 | Y_lsb
}
```
right?

Or maybe:
```cpp
Wire.requestFrom(address, (uint8_t)6);
if (6 <= Wire.available())
{
  X = float(int16_t(Wire.read() << 8 | Wire.read()));		//X_msb << 8 | X_lsb
  Z = float(int16_t(Wire.read() << 8 | Wire.read()));		//Z_msb << 8 | Z_lsb
  Y = float(int16_t(Wire.read() << 8 | Wire.read()));		//Y_msb << 8 | Y_lsb
}
```

**Comment by Koepel** [comment]
The `Wire.available()` with a while-loop and `micros()` will not make it safer. It is not doing something, and can be removed.

An integer can be 16 bits on a Arduino Uno, or 32 bits on a Arduino Zero or one of the MKR boards. The Wire.read() returns a `int`, which can be 16 or 32 bits. Since the sensor has always 16 bits, I prefer to build that 16-bits integer in the code. Only after that, a float can be made from it. The `float(int16_t(...` will probably also work, but I think my example is prettier to look at.

When 6 bytes are requested, the `Wire.available()` will not return a number larger than 6. Therefor the  `if (6 <= Wire.available())` is a little weird. It will work just fine, but I prefer `if (Wire.available() == 6)`.

There are a number of Wire libraries and compatible libraries. The most common Wire library is for the AVR microcontrollers like the Arduino Uno. Shortcutting SDA or SCL to GND will halt some functions of the AVR Wire library, and there is nothing you can do about that.

I'm not sure what you are trying to do with some code where you test the return value of `Wire.endTransmission()` after you collected data from that sensor.

**Comment by HeckRodSav** [comment]
Once I have the hardware in hand, I'll do the tests related to the changes you said, thank you!

About `Wire.endTransmission()`: I do many read/write in that part of the code, related to the BMP085 (barometer), and I need to ensure that the sensor stays connected throughout the process. Check if the transmission is still valid at the end it seemed enough.
Like:
```cpp
Wire.beginTransmission(address);

/* Many Wire.write(_)  */

if ( Wire.endTransmission() == 0) //So what I read is meaningful, I can consider it reliable
else //So it is not reliable, discard reading
```
In other situations, I use `Wire.endTransmission()` to reset a recalibration timer, if there is a long time between the last request and the last successful communication, it must recalibrate when it can connect, assuming the sensor has lost power due to the long time without communication (since the sensor is in one module).

**Comment by Koepel** [comment]
I'm not sure if your code can detect all problems. When, for example, the I2C is bad (a voltage mismatch of a 3.3V and 5V I2C bus or the I2C wires are too long), then a short check to see if the sensor is connected might be okay and a long I2C transaction with a lot of data might go wrong.

To check for a sensor at a certain I2C address, the `Wire.beginTransmission()` with `Wire.endTransmission()` is good.

The common way to determine if the returned data is okay is to check every return value of `Wire.endTransmission()` and every return value of `Wire.requestFrom()`. 
Instead of the return value of `Wire.requestFrom()` also the `Wire.available()` can be used to check if the same number of bytes was received as was requested.

If that is okay, then you can be sure that you have real data from the sensor.

To achieve that, reading and writing is often put into functions. Those functions can return `true` or `false` to indicate if it was successful.

**Comment by HeckRodSav** [comment]
Then at each `Wire.requestFrom()` I should check if the return is equivalent to the requested amount, okay. 

I wanted to make a system where I prepare all the sensors for reading:
```cpp
for(int i = 0; i < snsAmount; i++)
{
	Wire.beginTransmission(address[i]);
	Wire.write(value[i]);
	Wire.endTransmission();
}
```
And then I go through each one making the response request:
```cpp
for(int i = 0; i < snsAmount; i++)
{
	Wire.requestFrom(address[i], amount[i]);
	/* every process needed. */
}
```
Do you think it is viable? (The aim is to save time, making each sensor prepare simultaneously)

**Comment by Koepel** [comment]
The I2C transaction takes time and the `Wire.endTransmission()` and `Wire.requestFrom()` wait until the I2C transaction on the bus has finished. That means you can make extra functions and check returns values and so on. It does not matter a lot, since most time is spent in those function waiting for the I2C transaction to finish.

If the sensor itself needs time to prepare, then you can do two of those 'for'-statements. However, such sensors are very rare, so I suggest to keep it straightforward without fancy things.

Less rare are sensors that require that a Repeated Start **must** be used between setting the register address and reading data. If you want to use others sensors in the future, sooner or later you bump into a sensor that does not allow those two 'for'-statements.

**Comment by HeckRodSav** [comment]
My barometer needs that extra time, so it's worth use those `for`-statements. Since each sensor spends a few microseconds getting ready, if it is already done, better for the system.

This system is the embedded electronics of a rocket, the faster, the better.

I'll consider every detail of what you said in the next iteration.
Thank you so much!
